substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -383,14 +420,163 @@ public void setNativeLibs(NativeLibraries nativeLibs) {
+        // then inserting the name serves to allocate the short name associated with the inserted
+        // element's result index.
+        List<LookupName> bindings;
+
+        /**
+         * A lookup name is used as a key to record and subsequently lookup a short symbol that can
+         * replace one or more elements of a mangled name. A lookup name is usually just a simple
+         * string, i.e. some text that is to be mangled and to which the corresponding short symbol
+         * should decode. However, substitutions for namespace prefixes (e.g. AppLoader506::Foo)
+         * require a composite lookup name that composes a namespace prefix (AppLoader506) with a
+         * trailing simple name (Foo). The corresponding short symbol will demangle to the string
+         * produced by composing the prefix and simple name with a :: separator (i.e. restoring
+         * AppLoader506::Foo). Short symbols for pointer types (e.g. Foo* or AppLoader506::Foo*)
+         * require a pointer lookup name that identifies the substituted text as a pointer to some
+         * underlying type (e.g. Foo or AppLoader506::Foo). The corresponding short symbol will
+         * demangle to the string produced by demangling the underlying type with a * suffix. In
+         * theory the prefix for a pointer type lookup name might be defined by another pointer type
+         * lookup name (if, say, we needed to encode type Foo**). In practice, that case should not
+         * arise with Java method signatures.
+         */
+        private abstract class LookupName {
+            String value;
+
+            protected LookupName(String value) {
+                assert value != null;
+                this.value = value;
+            }
+
+            @Override
+            public abstract String toString();
+        }
+
+        private class SimpleLookupName extends LookupName {
+            SimpleLookupName(String value) {
+                super(value);
+            }
+
+            @Override
+            public boolean equals(Object other) {
+                if (other instanceof SimpleLookupName) {
+                    return this.value.equals(((SimpleLookupName) other).value);
+                }
+                return false;
+            }
+
+            @Override
+            public int hashCode() {
+                return value.hashCode();
+            }
+
+            @Override
+            public String toString() {
+                return value;
+            }
+        }
+
+        private abstract class CompositeLookupName extends LookupName {
+            LookupName tail;
+
+            CompositeLookupName(String value, LookupName tail) {
+                super(value);
+                assert tail != null;
+                this.tail = tail;
+            }
+
+            public boolean equals(Object other) {
+                if (other instanceof NamespaceLookupName) {
```

Comment:
Shouldn't this be `CompositeLookupName`?

```suggestion
                if (other instanceof CompositeLookupName) {
```

Reply:
Well caught!

Reply:
Done